docs/standard/serialization/snippets/system-text-json-converters-how-to/csharp/EnumConverter.cs

```diff
@@ -0,0 +1,206 @@
+                {
+                    Array values = Enum.GetValues(typeof(T));
+
+                    for (int i = 0; i < values.Length; i++)
+                    {
+                        T value = (T)values.GetValue(i)!;
+
+                        if (!TryProcessValue(value))
+                        {
+                            break;
+                        }
+                    }
+                }
+
+                bool TryProcessValue(T value)
+                {
+                    if (_readCache.Count == NameCacheLimit)
+                    {
+                        Debug.Assert(_writeCache.Count == NameCacheLimit);
+                        return false;
+                    }
+
+                    FormatAndAddToCaches(value, options.Encoder);
+                    return true;
+                }
+            }
+
+            public override T Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+            {
+                string? json;
+
+                if (reader.TokenType != JsonTokenType.String || (json = reader.GetString()) == null || !_readCache.TryGetValue(json, out T value))
+                {
+                    throw new JsonException();
+                }
+
+                return value;
+            }
+
+            public override void Write(
+                Utf8JsonWriter writer, 
+                T value, 
+                JsonSerializerOptions options)
+            {
+                if (!_writeCache.TryGetValue(value, out JsonEncodedText formatted))
+                {
+                    if (_writeCache.Count == NameCacheLimit)
+                    {
+                        Debug.Assert(_readCache.Count == NameCacheLimit);
+                        throw new ArgumentOutOfRangeException(
+                            "_writeCache.Count == NameCacheLimit", "");
+                    }
+
+                    formatted = FormatAndAddToCaches(value, options.Encoder);
+                }
+
+                writer.WriteStringValue(formatted);
+            }
+
+            private JsonEncodedText FormatAndAddToCaches(T value, JavaScriptEncoder? encoder)
+            {
+                (string valueFormattedToStr, JsonEncodedText valueEncoded) = FormatEnumValue(value.ToString(), _namingPolicy, encoder);
+                _readCache[valueFormattedToStr] = value;
+                _writeCache[value] = valueEncoded;
+                return valueEncoded;
+            }
+
+            private ValueTuple<string, JsonEncodedText> FormatEnumValue(string value, JsonNamingPolicy namingPolicy, JavaScriptEncoder? encoder)
+            {
+                string converted;
+
+                if (!value.Contains(ValueSeparator))
+                {
+                    converted = namingPolicy.ConvertName(value);
+                }
+                else
+                {
+                    // todo: optimize implementation here by leveraging https://github.com/dotnet/runtime/issues/934.
```

Comment:
Do you want to hold up this PR for those changes, or make those enhancements in another PR?

Reply:
I'm changing the PR to link to the issue that provides the code.